include/nbl/system/ISystem.h

```diff
@@ -0,0 +1,244 @@
+#ifndef __NBL_I_SYSTEM_H_INCLUDED__
+#define __NBL_I_SYSTEM_H_INCLUDED__
+
+#include <variant>
+#include "nbl/core/IReferenceCounted.h"
+#include "nbl/system/ISkippableAsyncQueueDispatcher.h"
+#include "nbl/system/IFileArchive.h"
+#include "nbl/system/IWriteFile.h"
+
+namespace nbl {
+namespace system
+{
+
+class ISystem final : public core::IReferenceCounted
+{
+public:
+    class ISystemCaller : core::IReferenceCounted
+    {
+    protected:
+        virtual ~ISystemCaller() = default;
+
+    public:
+        virtual core::smart_refctd_ptr<IReadFile> createReadFile(ISystem* sys, const std::filesystem::path& filename) = 0;
+        virtual core::smart_refctd_ptr<IWriteFile> createWriteFile(ISystem* sys, const std::filesystem::path& filename) = 0;
+        virtual uint32_t read(IReadFile* file, void* buffer, size_t offset, size_t size) = 0;
+        virtual uint32_t write(IWriteFile* file, const void* buffer, size_t offset, size_t size) = 0;
+    };
+
+protected:
+    struct SRequestType;
+    class CAsyncQueue;
+
+    class SFutureBase
+    {
+        friend ISystem;
+
+    protected:
+        bool valid_flag = false;
+        SRequestType* request;
+        CAsyncQueue* asyncQ;
+
+    public:
+        // SFuture is non-copyable and non-movable
+        SFutureBase(const SFutureBase&) = delete;
+
+        ~SFutureBase()
+        {
+            request->set_skip();
+        }
+
+        bool ready() const { return !request || request->ready; }
+        bool valid() const { return valid_flag; }
+
+        void wait()
+        {
+            asyncQ->waitForRequestCompletion(request[0]);
```

Comment:
A) in C++20 you'd be able to wait on the atomic_flag itself
B) the request itself should have its own lock+condvar, its kinda lame to have everybody contend on the same lock even if they'll unlock right away and go to sleep waiting for the condvar

Reply:
TL;DR maybe we dont need to keep a pointer to the queue

Reply:
> TL;DR maybe we dont need to keep a pointer to the queue

yep, we wouldnt have if each request had its own mutex
global mutex is still needed though for overflow cases and for making increment-with-wrap-around atomic

Reply:
> global mutex is still needed though for overflow cases and for making increment-with-wrap-around atomic

I think I put forward a solution that does not need a global mutex for overflow nor increment-with-wrap-around

Reply:
resolved.